### DISCUSSION TITLE
Notepad (scientific programming) style GUI

### DISCUSSION THREAD

**Discussion by abegertler** [Qalculate/qalculate-qt#175] [Ideas]
Hello!

Thank you so much for creating this. It is truly an increible project!

The one feature that keeps it from being truly useful for scientists or engineers making longer calculations is the ability to create a "script".

Other calculators, for example Calca.io, SpeQ, or Qaltex are good examples of this:
![image](https://github.com/user-attachments/assets/d17ec6bf-c1cb-4ecf-afe5-a7fa7141b707)
(As you can see for example in this Qaltex screenshot, the interface is beautiful. But it doesn't actually handle physical unit manipulations.)

![image](https://github.com/user-attachments/assets/c6f174c7-a6c6-46cc-aecb-49f3af5657b1)
(Calca.io handles units wonderfully (most of the time), and encounters errors often. The interface is very flexible (but not as beautifully arranged as Qaltex) and just like a notepad with markdown. Lacks a dark mode. Lacks permanent user units.)

For people making longer calculations, it makes much more sense to calculate from top to bottom, storing temporary variables and expressions and functions in a "file" type format, as you do when programming.
This is how matlab or python are used for scientific programming, but these are very expensive, clunkier, and don't handle physical units nearly as well as a program like yours!
A modern program like yours, which can handle scientific programming scripts, has the potential to be used by tens of millions of scientists, engineers, and students.

All of these above "notepad" style calculators listed above however, have their issues: Qaltex doesn't handle physical unit conversions well, SpeQ is waiting for some modernisation, and Calca is buggy and no longer supported.

Very curious what you think. Thank you for your consideration!

**Comment by hanna-kn** [comment]
This would probably be best implemented in a separate GUI, but I do not rule out the possibility of adding an alternative combined input and output widget to the Qt interface (the current ability to save different workspaces would be useful for this).

The terminal application, qalc, is currently the user interface (which is part of the project) closest to this, with top to bottom calculations, and support for reading expressions from a text file (this is also possible for the GUIs from the command line). It is however not possible to go back and edit previous expressions in interactive mode.

Cantor (https://cantor.kde.org/) can use libqalculate as backend.

**Comment by abegertler** [comment]
Hello Hanna,

Thank you so much for replying to me!! I am honored 😊

Right, this does feel like a separate GUI, as the workflow is quite different.
Hopefully it might be possible to mostly adapt the current Qt interface though.

I would feel remiss not to take the moment to re-iterate that I do think a notepad implementation of Qalculate is a tool that could replace excel and matlab and python for tens of millions of engineers. And the more I work with different physical units based calculators, the more I see that Qalculate is the closest to doing this.

And Cantor is a great lead! And looks wonderful in many ways.
(Unfortunately Cantor is not very well supported for windows, and lacks many of the features possible in the dedicated Qalculate application).

Thank you so much for developing Qalculate in any case!
Abe
